cs3api4lab/cs3apismanager.py

```diff
@@ -57,15 +143,379 @@ def save(self, model, path):
+
+        # ToDo: Implements run_pre_save_hook and run_post_save_hook
+        # self.run_pre_save_hook(model=model, path=path)
+
+        try:
+            if model['type'] == 'notebook':
+
+                nb = nbformat.from_dict(model['content'])
+                self.check_and_sign(nb, path)
+                self._save_notebook(path, nb)
+
+                # ToDo: Implements save to checkpoint
+                # if not self.checkpoints.list_checkpoints(path):
+                #     self.create_checkpoint(path)
+
+            elif model['type'] == 'file':
+                self._save_file(path, model['content'], model['format'])
+
+            elif model['type'] == 'directory':
+                self._save_directory(path)
+
+            else:
+                raise web.HTTPError(400, "Unhandled contents type: %s" % model['type'])
+
+        except web.HTTPError:
+            raise
+
+        except Exception as e:
+            self.log.error(u'Error while saving file: %s %s', path, e, exc_info=True)
+            raise web.HTTPError(500, u'Unexpected error while saving file: %s %s' % (path, e))
+
+        validation_message = None
+
+        if model['type'] == 'notebook':
+            self.validate_notebook_model(model)
+            validation_message = model.get('message', None)
+            model = self._notebook_model(path, content=False)
+
+        elif model['type'] == 'file':
+            model = self._file_model(path, content=False, format=None)
+
+        elif model['type'] == 'directory':
+            model = self._dir_model(path, content=False)
+
+        if validation_message:
+            model['message'] = validation_message
+
+        # self.run_post_save_hook(model=model, os_path=path)
+
+        return model
 
     def delete_file(self, path):
         """Delete the file or directory at path."""
-        raise NotImplementedError('cs3: missing')
+
+        path = self._normalize_path(path)
+
+        try:
+            self.cs3_file_api.remove(path, self.cs3_user_id, self.cs3_endpoint)
+
+        except FileNotFoundError as e:
+            self.log.error(u'Unknown error delete file: %s %s', path, e, exc_info=True)
+            raise web.HTTPError(404, u'No such file or directory: %s %s' % (path, e))
+
+        except Exception as e:
+            self.log.error(u'Unknown error delete file: %s %s', path, e, exc_info=True)
+            raise web.HTTPError(500, u'Unknown error delete file: %s %s' % (path, e))
 
     def rename_file(self, old_path, new_path):
         """Rename a file or directory."""
-        raise NotImplementedError('cs3: missing')
 
-    
+        if new_path == old_path:
+            return
+
+        #
+        # ToDo: Implements validate file like: notebook/services/contents/filemanager.py:587 using Reva API
+        #
+        old_path = self._normalize_path(old_path)
+        new_path = self._normalize_path(new_path)
+
+        # Should we proceed with the move?
+        if self.exists(new_path):
```

Comment:
Do we need an extra roundtrip for this? If the cs3apis don't allow us to force or disable a replace, then we should propose it. It would make it more efficient.

Reply:
If the file exists Reva created an error "error moving file", users don't know if the file exists or happen something else. Now it's better if we check it's file exists and throw a specific error.

Reply:
ok!